### PR TITLE
implement `setproperty!` for modules

### PR DESCRIPTION
**Pull request by simeonschaub**
It doesn't seem like there was any good reason for disallowing this.
Especially now with #43671 merged I found myself wanting this.

### CODE REVIEW

File: src/builtins.c [hunk 1]
```diff
@@ -940,22 +940,31 @@ JL_CALLABLE(jl_f_setfield)
...
-    jl_value_t *ft = jl_field_type_concrete(st, idx);
-    if (!jl_isa(args[2], ft))
-        jl_type_error("setfield!", ft, args[2]);
-    if (order >= jl_memory_order_acq_rel || order == jl_memory_order_release)
-        jl_fence(); // `st->[idx]` will have at least relaxed ordering
-    set_nth_field(st, v, idx, args[2], isatomic);
+    if (st == jl_module_type) {
+        JL_TYPECHK(setfield!, symbol, args[1]);
+        if (order != jl_memory_order_notatomic)
+            jl_atomic_error("setfield!: module binding cannot be written atomically");
```
Comment (vtjnash) [inline src/builtins.c:951]: It may be a lot more accurate to say that they can only be written atomically (just copy the fence logic below).

**Review by vtjnash — COMMENTED** [review]
LGTM, though we probably will want codegen support, and to implement swapproperty!/modifyproperty!/replaceproperty! also

**Review by vtjnash — COMMENTED** [review]
We might want to improve this text error to say "imported variable" or something similar:
```
julia> stdout; Core.setfield!(Main, :stdout, stdout)
stdout = Base.TTY(RawFD(13) open, 0 bytes waiting)
ERROR: cannot assign a value to variable Base.stdout from module Main
Stacktrace:
 [1] top-level scope
   @ REPL[1]:1
```

File: src/builtins.c (hunk 1 above)
Comment (simeonschaub) [reply src/builtins.c:951]: Do you mean the fence should be the default? I think that would be inconsistent with regular assignment to a global, which uses a relaxed memory order.

**Comment by simeonschaub** [comment]
Ok, I can look into what happens in codegen

File: src/builtins.c (hunk 1 above)
Comment (vtjnash) [reply src/builtins.c:951]: The `order` should be respected from the user.

```suggestion
        if (order == jl_memory_order_notatomic)
            jl_atomic_error("setfield!: module binding cannot be written non-atomically");
```

(we should perhaps change all "cannot be written non-atomically" to "must be written atomically" also, to word them in the positive)

Comment (simeonschaub) [reply src/builtins.c:951]: I thought our `not_atomic` still corresponded to a unordered/relaxed memory model? Or is that wrong?

**Review by vtjnash — APPROVED** [review]

**Comment by simeonschaub** [comment]
I think I want to tackle codegen in another PR. This should ideally be consistent with regular assignment to a global, where we currently just call into `jl_checked_assignment`, so we should ideally teach LLVM to reason about both ways to assign to globals.

**Comment by aviatesk** [comment]
What will happen if we call `setfield(M, :var, v)` where `M.var` is a constant value?

**Comment by simeonschaub** [comment]
It ends up calling `jl_checked_assignement`, so the behavior should be the same as with regular assignments to globals.

**Review by aviatesk — APPROVED** [review]
LGTM.

File: src/builtins.c [hunk 2]
```diff
@@ -940,22 +940,33 @@ JL_CALLABLE(jl_f_setfield)
...
-    if (isatomic == (order == jl_memory_order_notatomic))
-        jl_atomic_error(isatomic ? "setfield!: atomic field cannot be written non-atomically"
-                                 : "setfield!: non-atomic field cannot be written atomically");
-    jl_value_t *ft = jl_field_type_concrete(st, idx);
-    if (!jl_isa(args[2], ft))
-        jl_type_error("setfield!", ft, args[2]);
-    if (order >= jl_memory_order_acq_rel || order == jl_memory_order_release)
-        jl_fence(); // `st->[idx]` will have at least relaxed ordering
-    set_nth_field(st, v, idx, args[2], isatomic);
+    if (st == jl_module_type) {
```
Comment (Keno) [inline src/builtins.c:948]: In recent discussions, @JeffBezanson considered `getfield` on modules to be a mistake and preferred a separate builtin instead for 2.0. Maybe we should avoid repeating this for setfield! and just go straight for a separate builtin?

Comment (simeonschaub) [reply src/builtins.c:948]: Maybe, but I think with the current `getfield` that would arguably be more of a special case, no? I'd be happy to change that though.

Comment (Keno) [reply src/builtins.c:948]: @JeffBezanson should say which he prefers. I had suggested implementing the `getbinding` builtin right away also and just keeping getfield working with the intent to deprecate in 2.0. That way there wouldn't be an asymmetry and `getfield` would just have a bit of a legacy quirk.

**Comment by simeonschaub** [comment]
Closing in favor of #44231

**Pull request closed without merging**